clang/lib/Sema/BoundsAnalysis.cpp

```diff
@@ -0,0 +1,220 @@
+          DeclMap[D]++;
+      }
+    }
+  }
+  EB->Gen = Intersections;
+
+  // Now update EB->Gen.
+  for (const auto item : DeclMap) {
+    unsigned count = item.second;
+    if (count == EB->Block->pred_size()) {
+      const auto *D = item.first;
+      if (!EB->Gen.count(D))
+        EB->Gen[D] = 1;
+      else
+        EB->Gen[D]++;
+    }
+  }
+}
+
+void BoundsAnalysis::UpdateInMap(ElevatedCFGBlock *EB, BlockMapTy BlockMap) {
+  // In(B) = min(n Out(B')), where B E preds(B).
+
+  // In(B) is the intersection of the Out's of all preds of B. For the
+  // intersection, we pick the Decl with the smaller upper bound. For example,
+  // if preds(B) = {X, Y} and Out(X)= {p:1, q:2, r:0} and Out(Y) = {p:0, q:3,
+  // s:2} then In(B) = {p:0, q:2}.
+  BoundsMap Intersections;
+  bool ItersectionEmpty = true;
+
+  for (const CFGBlock *pred : EB->Block->preds()) {
+    auto PredEB = BlockMap[pred];
+    if (ItersectionEmpty) {
+      Intersections = PredEB->Out;
+      ItersectionEmpty = false;
+    } else
+      Intersections = Intersect(Intersections, PredEB->Out);
+  }
+  EB->In = Intersections;
+}
+
+BoundsMap BoundsAnalysis::UpdateOutMap(ElevatedCFGBlock *EB) {
+  // Out(B) = max(Gen(B) u In(B)).
+
+  // Out(B) is the union of In(B) and Gen(B). If both In and Gen have the same
+  // Decl then for the union we pick the one with the larger upper bound. For
+  // example, if In(B) = {p:1, q:2} and Gen(B) = {p:2}, then Out(B) = {p:2,
+  // q:2}.
+  auto OldOut = EB->Out;
+  EB->Out = Union(EB->In, EB->Gen);
+  return OldOut;
+}
+
+void BoundsAnalysis::CollectWidenedBounds(BlockMapTy BlockMap) {
+  for (auto item : BlockMap) {
+    const auto *B = item.first;
+    auto *EB = item.second;
+    WidenedBounds[B] = EB->Out;
+  }
+}
+
+BoundsMap BoundsAnalysis::GetWidenedBounds(const CFGBlock *B) {
+  return WidenedBounds[B];
+}
+
+const Expr *BoundsAnalysis::GetTerminatorCondition(const CFGBlock *B) const {
+  // TODO: Handle other Stmt types, like while loops, etc.
+  if (const Stmt *S = B->getTerminator())
+    if (const auto *IfS = dyn_cast<IfStmt>(S))
+      return IfS->getCond();
+  return nullptr;
+}
+
+const VarDecl *BoundsAnalysis::GetVarDecl(const Expr *E) const {
+  if (const auto *CE = dyn_cast<CastExpr>(E))
+    if (const auto *UO = dyn_cast<UnaryOperator>(CE->getSubExpr()))
+      if (auto *D = dyn_cast<DeclRefExpr>(UO->getSubExpr()->IgnoreImplicit()))
+        return dyn_cast<VarDecl>(D->getDecl());
+  return nullptr;
+}
+
+bool BoundsAnalysis::IsPointerDerefLValue(const Expr *E) const {
+  if (const auto *UO = dyn_cast<UnaryOperator>(E))
+    return UO->getOpcode() == UO_Deref;
+  return false;
+}
+
+bool BoundsAnalysis::ContainsPointerDeref(const Expr *E) const {
+  if (const auto *CE = dyn_cast<CastExpr>(E)) {
+    if (CE->getCastKind() != CastKind::CK_LValueToRValue)
+      return false;
+    return IsPointerDerefLValue(CE->getSubExpr());
+  }
+  return false;
+}
+
+BoundsMap BoundsAnalysis::Intersect(BoundsMap &A, BoundsMap &B) {
```

Comment:
Maybe add a comment that "Intersect" mutates the first argument. Or make the second argument of "intersect", "union", "differ" to be const?